File: src/main/java/com/ptah/auth/WithAuthorityAspect.java
package com.ptah.auth;

import org.aspectj.lang.JoinPoint;
import org.aspectj.lang.annotation.Aspect;
import org.aspectj.lang.annotation.Before;
import org.aspectj.lang.annotation.Pointcut;
import org.springframework.security.core.Authentication;
import org.springframework.security.core.context.SecurityContextHolder;
import org.springframework.security.core.GrantedAuthority;
import org.springframework.stereotype.Component;

@Aspect
@Component
public class WithAuthorityAspect {

    @Pointcut("@annotation(withAuthority)")
    public void hasPermissionInterceptor(WithAuthority withAuthority) {
    }

    @Before(value = "hasAuthorityInterceptor(withAuthority)", argNames = "joinPoint,withAuthority")
    public void hasPermissionCheck(JoinPoint joinPoint, WithAuthority withPermission) {

        Authentication auth = SecurityContextHolder.getContext().getAuthentication();
        if (auth == null) {
            throw new SecurityException("Authentication not found.");
        }

        boolean hasPermission = auth.getAuthorities().stream()
                .map(GrantedAuthority::getAuthority)
                .anyMatch(role -> role.equals(withPermission.value()));

        if (!hasPermission) {
            throw new SecurityException("Insufficient permissions for method invocation.");
        }
    }
}

File: src/main/kotlin/com/ptah/auth/WithAuthorityAspect.kt
package com.ptah.auth

import org.aspectj.lang.JoinPoint
import org.aspectj.lang.annotation.Aspect
import org.aspectj.lang.annotation.Before
import org.aspectj.lang.annotation.Pointcut
import org.springframework.security.core.GrantedAuthority
import org.springframework.security.core.context.SecurityContextHolder
import org.springframework.stereotype.Component

@Aspect
@Component
class WithAuthorityAspect {
    @Pointcut("@annotation(withAuthority)")
    fun hasAuthorityInterceptor(withAuthority: WithAuthority?) {
    }

    @Before(value = "hasAuthorityInterceptor(withAuthority)", argNames = "joinPoint,withAuthority")
    fun hasPermissionCheck(joinPoint: JoinPoint?, withPermission: WithAuthority) {
        val auth = SecurityContextHolder.getContext().authentication
            ?: throw SecurityException("Authentication not found.")
        val hasPermission = auth.authorities.stream()
            .map { obj: GrantedAuthority -> obj.authority }
            .anyMatch { role: String -> role == withPermission.value }
        if (!hasPermission) {
            throw SecurityException("Insufficient permissions for method invocation.")
        }
    }
}
